test/export/test_export.py

```diff
@@ -5048,6 +5048,8 @@ def forward(self, x):
         export(f, (inputs,), dynamic_shapes=dynamic_shapes)
 
     def test_disable_forced_specializations(self):
+        return  # TODO: reenable
```

Comment:
@avikchaudhuri Need advice on these two tests.

Failure looks like this:

```
======================================================================
FAIL: test_disable_forced_specializations_errors (__main__.TestExport)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/data/users/ezyang/a/pytorch/torch/export/_trace.py", line 1335, in _non_strict_export
    produce_guards_and_solve_constraints(
  File "/data/users/ezyang/a/pytorch/torch/_export/non_strict_utils.py", line 304, in produce_guards_and_solve_constraints
    raise constraint_violation_error
  File "/data/users/ezyang/a/pytorch/torch/_export/non_strict_utils.py", line 265, in produce_guards_and_solve_constraints
    shape_env.produce_guards(
  File "/data/users/ezyang/a/pytorch/torch/fx/experimental/symbolic_shapes.py", line 4159, in produce_guards
    raise ConstraintViolationError(
torch.fx.experimental.symbolic_shapes.ConstraintViolationError: Constraints violated (dx0, dz)! For more information, run with TORCH_LOGS="+dynamic".
  - The values of dz = L['args'][0][3].size()[0] and dy = L['args'][0][2].size()[0] must always be equal.
  - Not all values of dx0 = L['args'][0][1].size()[0] in the specified range dx0 <= 9223372036854775806 satisfy the generated guard 4 <= L['args'][0][1].size()[0] and L['args'][0][1].size()[0] <= 9223372036854775806
Suggested fixes:
  dz = dy

During handling of the above exception, another exception occurred:

torch._dynamo.exc.UserError: Constraints violated (dx0, dz)! For more information, run with TORCH_LOGS="+dynamic".
  - The values of dz = L['args'][0][3].size()[0] and dy = L['args'][0][2].size()[0] must always be equal.
  - Not all values of dx0 = L['args'][0][1].size()[0] in the specified range dx0 <= 9223372036854775806 satisfy the generated guard 4 <= L['args'][0][1].size()[0] and L['args'][0][1].size()[0] <= 9223372036854775806
Suggested fixes:
  dz = dy

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/data/users/ezyang/a/pytorch/torch/testing/_internal/common_utils.py", line 2715, in wrapper
    method(*args, **kwargs)
  File "/data/users/ezyang/a/pytorch/test/export/test_export.py", line 5213, in test_disable_forced_specializations_errors
    with self.assertRaisesRegex(  # if allow = False, suggested fixes should specialize 3, 4, 12.
AssertionError: ".*Specializations unexpectedly required(.*\n)*Suggested fixes:(.*\n)*.*dw0 = 3(.*\n)*.*dw1 = 4(.*\n)*.*dx0 = 12(.*\n)*.*dz = dy(.*\n)*" does not match "Constraints violated (dx0, dz)! For more information, run with TORCH_LOGS="+dynamic".
  - The values of dz = L['args'][0][3].size()[0] and dy = L['args'][0][2].size()[0] must always be equal.
  - Not all values of dx0 = L['args'][0][1].size()[0] in the specified range dx0 <= 9223372036854775806 satisfy the generated guard 4 <= L['args'][0][1].size()[0] and L['args'][0][1].size()[0] <= 9223372036854775806
Suggested fixes:
  dz = dy"

To execute this test, run the following from the base repo dir:
     python test/export/test_export.py -k TestExport.test_disable_forced_specializations_errors

This message can be suppressed by setting PYTORCH_PRINT_REPRO_ON_FAILURE=0
```

What seems to have happened is that, previously, there were enough guards that it caused a forced specialization. I've perturbed the guards a bit, in particular, we sometimes generate `x <= 9223372036854775806` guards instead of whatever guards we were doing before. It seems to me that we are now no longer exercising the forced specialization path.

I am reasonably confident that the new guards are correct. In particular, previously, if we had an int input where we did something special when given sys.maxsize, we now properly guard in this case (whereas previously we would have dropped this entirely). But I do not know how to best adjust this test to keep it continuing to test what it did before.

Reply:
@pianpwk can you look at this?
The only part of this test that might involve a `dz` is the 2d reshape I think?
```
class FreeReshape(torch.nn.Module):
            def forward(self, x, y, z):
                return x.reshape([-1]) + y.reshape([-1]) + z  # s0*s1 = s2*s3 = s4
```

But this shouldn't produce `dz = dy` and moreover the dim names are created with suffixes in the test so I'm puzzled why `dz` and `dy` are showing up in the error. Am I looking at the right test, @ezyang ?

Could it be https://github.com/pytorch/pytorch/blob/1c058c5634bcd73d132a52db4a8d1109646a2489/test/export/test_export.py#L5198 instead?

Then `dz = dy` would be fine. I'm not sure what's happening with `dx0`, does the range get derived here?

Reply:
Actually `dx0` is just getting a new lower bound here of `4` because of the reshape. That looks legit too. So the real question is what happened to all the forced specializations lol. In other words, where did the complex guards disappear? (Did sympy magically simplify them?)

Reply:
Oh interesting, I ran with TORCH_LOGS="+dynamic" and I see this line:
`skipped set_replacement s4 = s0*s1 (solve) [VR[4, 85070591730234615828950163710522949636] not subset of VR[4, 9223372036854775806]]`

this is probably why the complex guard disappears

Reply:
And yeah, the error log posted about for `test_disable_forced_specializations_errors` seems to be the same thing, it's bypassing the complex guard s0*s1 = s2 and hitting the dz = dy suggested fix instead.

Reply:
Also I don't know if it's out of scope for this PR, but I guess we should be using this in torch/export/dynamic_shapes.py? I still see sys.maxsize related issues for examples like 2*_dx. Probablyyyy this is why the is_subset() call acts weird for export?

```
def test_mul(self):
    class Foo(torch.nn.Module):
        def forward(self, x):
            return x.reshape([-1, 2])
    inputs = (torch.randn(8),)
    _dx = Dim("_dx")
    ep = export(Foo(), inputs, dynamic_shapes={"x": (2*_dx,)})

======================================================================
FAIL: test_mul (__main__.TestExport)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/data/users/pianpwk/pytorch/torch/testing/_internal/common_utils.py", line 2719, in wrapper
    method(*args, **kwargs)
  File "/data/users/pianpwk/pytorch/test/export/test_export.py", line 5065, in test_mul
    ep = export(Foo(), inputs, dynamic_shapes={"x": (2*_dx,)})
  File "/data/users/pianpwk/pytorch/torch/export/__init__.py", line 174, in export
    return _export(
  File "/data/users/pianpwk/pytorch/torch/export/_trace.py", line 944, in wrapper
    raise e
  File "/data/users/pianpwk/pytorch/torch/export/_trace.py", line 927, in wrapper
    ep = fn(*args, **kwargs)
  File "/data/users/pianpwk/pytorch/torch/export/exported_program.py", line 88, in wrapper
    return fn(*args, **kwargs)
  File "/data/users/pianpwk/pytorch/torch/export/_trace.py", line 1454, in _export
    aten_export_artifact = export_func(
  File "/data/users/pianpwk/pytorch/torch/export/_trace.py", line 1059, in _strict_export
    gm_torch_level = _export_to_torch_ir(
  File "/data/users/pianpwk/pytorch/torch/export/_trace.py", line 511, in _export_to_torch_ir
    gm_torch_level, _ = torch._dynamo.export(
  File "/data/users/pianpwk/pytorch/torch/_dynamo/eval_frame.py", line 1253, in inner
    constraints = _process_dynamic_shapes(_f, args, kwargs, dynamic_shapes)
  File "/data/users/pianpwk/pytorch/torch/export/dynamic_shapes.py", line 871, in _process_dynamic_shapes
    assoc_shapes(combined_args, dynamic_shapes)
  File "/data/users/pianpwk/pytorch/torch/export/dynamic_shapes.py", line 863, in assoc_shapes
    _tree_map(assoc_shape, combined_args, dynamic_shapes)
  File "/data/users/pianpwk/pytorch/torch/export/dynamic_shapes.py", line 603, in _tree_map
    return tree_map(f, tree, *dynamic_shapes, is_leaf=is_leaf)
  File "/data/users/pianpwk/pytorch/torch/utils/_pytree.py", line 948, in tree_map
    return treespec.unflatten(map(func, *flat_args))
  File "/data/users/pianpwk/pytorch/torch/utils/_pytree.py", line 787, in unflatten
    leaves = list(leaves)
  File "/data/users/pianpwk/pytorch/torch/export/dynamic_shapes.py", line 601, in f
    return func(t, *dynamic_shapes)
  File "/data/users/pianpwk/pytorch/torch/export/dynamic_shapes.py", line 854, in assoc_shape
    update_symbols(t, dynamic_shape)
  File "/data/users/pianpwk/pytorch/torch/export/dynamic_shapes.py", line 834, in update_symbols
    check_same_bounds(dim)
  File "/data/users/pianpwk/pytorch/torch/export/dynamic_shapes.py", line 808, in check_same_bounds
    bounds[dim.__name__] = (dim.min, dim.max)
  File "/data/users/pianpwk/pytorch/torch/export/dynamic_shapes.py", line 147, in max
    assert _max_symint <= sys.maxsize - 1, (
AssertionError: Expected derived max value of 2*_dx to be <= 9223372036854775806. Please specify an appropriate max value for _dx (currently 9223372036854775806).

To execute this test, run the following from the base repo dir:
     python test/export/test_export.py -k TestExport.test_mul

This message can be suppressed by setting PYTORCH_PRINT_REPRO_ON_FAILURE=0
```

Reply:
Anywhere you're using maxsize to bound above/below you can use int_oo. I probably wouldn't recommend using it in user visible API though.

Reply:
>  Am I looking at the right test, @ezyang ?

Sorry, @avikchaudhuri, both tests fail but they fail in different ways. I pasted the error message for the errors test. The non errors test fails with

```
E0607 06:26:14.084000 140011261854720 torch/_guards.py:261] [0/0] torch.fx.experimental.symbolic_shapes.ConstraintViolationError: Constraints violated (dx)! For more information, run with TORCH_LOGS="+dynamic".
E0607 06:26:14.084000 140011261854720 torch/_guards.py:261] [0/0]   - Not all values of dx = L['x'].size()[0] in the specified range dx <= 9223372036854775806 satisfy the generated guard Ne(Mod(L['x'].size()[0], L['x'].size()[0] - 1), 0).
E0607 06:26:14.086000 140011261854720 torch/_guards.py:263] [0/0] Created at:
E0607 06:26:14.086000 140011261854720 torch/_guards.py:263] [0/0]   File "/data/users/ezyang/a/pytorch/torch/_dynamo/convert_frame.py", line 563, in transform
E0607 06:26:14.086000 140011261854720 torch/_guards.py:263] [0/0]     tracer = InstructionTranslator(
E0607 06:26:14.086000 140011261854720 torch/_guards.py:263] [0/0]   File "/data/users/ezyang/a/pytorch/torch/_dynamo/symbolic_convert.py", line 2359, in __init__
E0607 06:26:14.086000 140011261854720 torch/_guards.py:263] [0/0]     output=OutputGraph(
E0607 06:26:14.086000 140011261854720 torch/_guards.py:263] [0/0]   File "/data/users/ezyang/a/pytorch/torch/_dynamo/output_graph.py", line 312, in __init__
E0607 06:26:14.086000 140011261854720 torch/_guards.py:263] [0/0]     self.init_ambient_guards()
E0607 06:26:14.086000 140011261854720 torch/_guards.py:263] [0/0]   File "/data/users/ezyang/a/pytorch/torch/_dynamo/output_graph.py", line 451, in init_ambient_guards
E0607 06:26:14.086000 140011261854720 torch/_guards.py:263] [0/0]     self.guards.add(ShapeEnvSource().make_guard(GuardBuilder.SHAPE_ENV))
stats [('calls_captured', 42), ('unique_graphs', 3)]
unimplemented [('Constraints violated (dx)! For more information, run with TORCH_LOGS="+dynamic".\n  - Not all values of dx = L[\'x\'].size()[0] in the specified range dx <= 9223372036854775806 satisfy the generated guard Ne(Mod(L[\'x\'].size()[0], L[\'x\'].size()[0] - 1), 0).\nSpecializations unexpectedly required (dx, dy)! For more information, run with TORCH_LOGS="+dynamic".\n  - dx = L[\'x\'].size()[0] must be specialized to 10 because the guards generated for it are too complex.\n  - dy = L[\'x\'].size()[1] must be specialized to 72 because the guards generated for it are too complex.\n\nSuggested fixes:\n  dx = 10\n  dy = 72', 1)]
F
======================================================================
FAIL: test_disable_forced_specializations_ok (__main__.TestExport)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/data/users/ezyang/a/pytorch/torch/testing/_internal/common_utils.py", line 2719, in wrapper
    method(*args, **kwargs)
  File "/data/users/ezyang/a/pytorch/test/export/test_export.py", line 5112, in test_disable_forced_specializations_ok
    with self.assertRaisesRegex(  # this will force specialize
AssertionError: UserError not raised

To execute this test, run the following from the base repo dir:
     python test/export/test_export.py -k TestExport.test_disable_forced_specializations_ok

This message can be suppressed by setting PYTORCH_PRINT_REPRO_ON_FAILURE=0

----------------------------------------------------------------------
```

Reply:
@pianpwk oh good day, consistently using int_oo instead of sys.maxsize in torch/export/dynamic_shapes.py works.